sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala

```diff
@@ -247,58 +260,130 @@ object CTESubstitution extends Rule[LogicalPlan] {
+        .map(_._2.copy(child = substituted, recursive = recursionFound))
+        .getOrElse(CTERelationDef(substituted))
       if (!alwaysInline) {
         cteDefs += cteRelation
       }
+
+      // From this point any reference to the definition is non-recursive.
+      val nonRecursiveCTERelation = if (cteRelation.recursive) {
+        cteRelation.copy(recursive = false)
+      } else {
+        cteRelation
+      }
+
       // Prepending new CTEs makes sure that those have higher priority over outer ones.
-      resolvedCTERelations +:= (name -> cteRelation)
+      resolvedCTERelations +:= (name -> nonRecursiveCTERelation)
     }
     resolvedCTERelations
   }
 
+  /**
+   * @return tuple of the logical plan and recursionFound boolean.
+   */
   private def resolveWithCTERelations(
       table: String,
       alwaysInline: Boolean,
       cteRelations: Seq[(String, CTERelationDef)],
-      unresolvedRelation: UnresolvedRelation): LogicalPlan = {
-    cteRelations
+      unresolvedRelation: UnresolvedRelation): (LogicalPlan, Boolean) = {
+    var recursionFound = false
+    val cteRelationsOut = cteRelations
       .find(r => conf.resolver(r._1, table))
       .map {
         case (_, d) =>
           if (alwaysInline) {
             d.child
           } else {
+            if (d.recursive) {
+              // self-reference is found
+              recursionFound = true
+            }
             // Add a `SubqueryAlias` for hint-resolving rules to match relation names.
-            SubqueryAlias(table, CTERelationRef(d.id, d.resolved, d.output, d.isStreaming))
+            SubqueryAlias(table,
+              CTERelationRef(d.id, d.resolved, d.output, d.isStreaming, recursive = d.recursive))
           }
       }
       .getOrElse(unresolvedRelation)
+    (cteRelationsOut, recursionFound)
   }
 
+  /**
+   * @return tuple of the logical plan and recursionFound boolean.
+   */
   private def substituteCTE(
       plan: LogicalPlan,
       alwaysInline: Boolean,
-      cteRelations: Seq[(String, CTERelationDef)]): LogicalPlan = {
-    plan.resolveOperatorsUpWithPruning(
+      cteRelations: Seq[(String, CTERelationDef)],
+      recursiveCTERelation: Option[(String, CTERelationDef)]): (LogicalPlan, Boolean) = {
+    var recursionFound = false
+    val substituted = plan.resolveOperatorsUpWithPruning(
         _.containsAnyPattern(RELATION_TIME_TRAVEL, UNRESOLVED_RELATION, PLAN_EXPRESSION,
           UNRESOLVED_IDENTIFIER)) {
       case RelationTimeTravel(UnresolvedRelation(Seq(table), _, _), _, _)
         if cteRelations.exists(r => plan.conf.resolver(r._1, table)) =>
         throw QueryCompilationErrors.timeTravelUnsupportedError(toSQLId(table))
 
       case u @ UnresolvedRelation(Seq(table), _, _) =>
-        resolveWithCTERelations(table, alwaysInline, cteRelations, u)
+        val resolved = resolveWithCTERelations(table, alwaysInline,
+          (recursiveCTERelation ++ cteRelations).toSeq, u)
```

Comment:
why not also make `resolveWithCTERelations` to take an optional parameter `recursiveCTERelation: Option[(String, CTERelationDef)]` like this function `substituteCTE` does? Then `CTERelationDef#recursive` is not needed.

Reply:
1. I will make `resolveWithCTERelations` to take `recursiveCTERelation` as you suggested.
-------
2. Regarding the question if we need `CTERelationDef#recursive`, I still see it necessary and there are following reasons for that:
- this flag is actually needed in `resolveWithCTERelation` to be passed down to `CTERelationRef` to have resolved if this is a self-reference or not. Maybe there are other ways to avoid it, but it seems to be the most elegant currently. Let me know if you have any other suggestions.
- it is also needed later in the PRs to come. You can check initial Peter's PR on recursive CTEs (https://github.com/apache/spark/pull/40744/files), specifically the file `InsertLoops.scala`, where it is decided depending on this flag if we transform this CTE to UnionLoop class (which will be introduced later) or not.
    - If we do not have this flag, we would need to run again and for every `CTERelationDef` function `substituteCTE` to see if this is a self-reference. 

Maybe I am wrong, but at least this is why I see `CTERelationDef#recursive` necessary as of now

Reply:
OK, then should we make `CTERelationDef#recursive` a lazy val which is computed by checking if there is any self reference in its child?

Reply:
This looks safer as we can avoid inconsistency issue completely. Otherwise, what happens if the `recursive` flag is true but there is no self reference in the child?

Reply:
> This looks safer as we can avoid inconsistency issue completely. Otherwise, what happens if the `recursive` flag is true but there is no self reference in the child?

I am thinking now if this is really possible to happen, since the only way for the `recursive` flag to become true is if `resolveWithCTERelations` sets `recursionFound=True`, no other way if I look at the code correctly

Reply:
> OK, then should we make `CTERelationDef#recursive` a lazy val which is computed by checking if there is any self reference in its child?

This is solved automatically only by prepanding `lazy` keyword in front of `recursive` or some more logic and consideration afterwards is necessary?

Reply:
`resolveWithCTERelations` is refactored now but seems like I have a bug. I will post here as soon as I solve it